api/controllers/login.go

```diff
@@ -0,0 +1,80 @@
+package controllers
+
+import (
+  "net/http"
+  "github.com/gin-gonic/gin"
+  "github.com/charmbracelet/log"
+  "github.com/Salz-Team/salz/api/models"
+  "strconv"
+
+  ghlib "github.com/google/go-github/v62/github"
+)
+
+// TODO find a good place for this
+const (
+  DEFAULT_ELO = 1000.0
+)
+
+func (ctrl *Controller) OAuthLoginHandler(c *gin.Context) {
+  url := ctrl.cfg.OAuth2Config.AuthCodeURL("ain't got no state yet")
+  log.Info("Login redirect", "url", url)
+  c.Redirect(http.StatusTemporaryRedirect, url)
+}
+
+func (ctrl *Controller) OAuthCallbackHandler(c *gin.Context) {
+  // TODO Check the state to prevent CSRF attacks
+  code := c.Query("code")
+  token, err := ctrl.cfg.OAuth2Config.Exchange(c, code)
+  if err != nil {
+    log.Error("oauth2: Exchange failed", "error", err)
+    c.AbortWithStatus(http.StatusUnauthorized)
```

Comment:
should it be a 500?

Reply:
Can you share the doc you followed while implementing this? I'll need to know which "code" is `c.Query("code")` to know if this is a problem between the SP and the IdP. If it is, then this would be a 500 series error. If you know that this code isn't exposed or generated by the user but by the api server for its own auth with the IdP, then this is a 500.

Given that the API doc for oauth2 don't seem to show what errors can be returned (and looking at the code, there are tons) though, I think we can probably stick with 500 and not decide whether it should be some other code from the 500s.

Reply:
code is returned by the identity provider once the user allows access. It's generated by the identity provider, and we exchange it for an access token: https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#2-users-are-redirected-back-to-your-site-by-github

This is not necessarily a problem with the identity provider. The code is valid for only a short amount of time, so using an expired code gives us an unauthorized response when trying to exchange. I'm not sure what other failure modes there are.

I think I asked if this should be a 500 error, because I feel like returning an unauthorized response to the user feels wrong if they literally just clicked "allow access".

Reply:
- not 500 because we can't do anything about it and error monitoring will freak out
- unauthorized also covers the case where someone tampers with the code and code exchange fails.